cgi-bin/DW/Controller/Inbox.pm

```diff
@@ -0,0 +1,659 @@
+        push @cleaned_items, $cleaned;
+
+    }
+
+    my $vars = {
+        messages => \@cleaned_items,
+        page     => $page,
+        view     => $view
+    };
+
+    return DW::Template->template_string( 'inbox/msg_list.tt', $vars );
+
+}
+
+sub render_folders {
+    my ( $remote, $view ) = @_;
+    my $user_messsaging = LJ::is_enabled('user_messaging');
+    my $inbox           = $remote->notification_inbox
+        or return error_ml( "/inbox/index.tt.error.couldnt_retrieve_inbox",
+        { 'user' => $remote->{user} } );
+
+    my $vars;
+
+    my @children = (
+        {
+            view     => 'circle',
+            label    => 'circle_updates',
+            unread   => $inbox->circle_event_count,
+            children => [
+                { view => 'birthday',  label => 'birthdays' },
+                { view => 'encircled', label => 'encircled' }
+            ]
+        },
+        {
+            view   => 'entrycomment',
+            label  => 'entries_and_comments',
+            unread => $inbox->entrycomment_event_count
+        },
+        { view => 'pollvote', label => 'poll_votes', unread => $inbox->pollvote_event_count },
+        {
+            view   => 'communitymembership',
+            label  => 'community_membership',
+            unread => $inbox->communitymembership_event_count
+        },
+        {
+            view   => 'sitenotices',
+            label  => 'site_notices',
+            unread => $inbox->sitenotices_event_count
+        }
+
+    );
+
+    if ($user_messsaging) {
+
+       # push links for recieved PMs and sent PMs to the beginning and end of the list, respectively
+        unshift @children,
+            {
+            view   => 'usermsg_recvd',
+            label  => 'messages',
+            unread => $inbox->usermsg_recvd_event_count
+            };
+        push @children,
+            { view => 'usermsg_sent', label => 'sent', unread => $inbox->usermsg_sent_event_count };
+    }
+    push @children, { view => 'bookmark', label => 'bookmarks', unread => $inbox->bookmark_count };
+    push @children, { view => 'archived', label => 'archive' } if LJ::is_enabled('esn_archive');
+
+    $vars->{folder_links} = {
+        view     => 'all',
+        label    => 'all',
+        unread   => $inbox->all_event_count,
+        children => \@children
+    };
+    $vars->{user_messaging} = $user_messsaging;
+    $vars->{view}           = $view;
+    return DW::Template->template_string( 'inbox/folders.tt', $vars );
+}
+
+sub action_handler {
+    my ( $ok, $rv ) = controller();
+    return $rv unless $ok;
+
+    # gets the request and args
+    my $r         = $rv->{r};
+    my $args      = $r->json;
+    my $action    = $args->{action};
+    my $ids       = $args->{'ids'};
+    my $view      = $args->{view} || 'all';
+    my $page      = $args->{page} || 1;
+    my $itemid    = $args->{itemid} || 0;
+    my $remote    = LJ::get_remote();
+    my $form_auth = $args->{lj_form_auth};
```

Comment:
This `$form_auth` seems to be unused so now I'm wondering if this needs to be checked in the `controller()` call above? I.e., if this handler is only invoked on POST, we probably need to validate the form auth?